Restart Camera During Main Descent

akoen opened the pull request.

jesseli2002 committed:
Initial

jesseli2002 committed:
Tests

jesseli2002 committed:
First proper task

jesseli2002 committed:
Factor out fakes

jesseli2002 committed:
Clock blocker

jesseli2002 committed:
Reorganize folders

jesseli2002 committed:
Repeating

jesseli2002 committed:
More tests

jesseli2002 committed:
Fix unscheduling bug

jesseli2002 committed:
More test

jesseli2002 committed:
Organize for library

jesseli2002 committed:
so apparently didn't actually include this

jesseli2002 committed:
get task

jesseli2002 commented on avionics/common/include/tasks/restart_camera.hpp:
```diff
@@ -0,0 +1,24 @@
+#pragma once
+#include "cameras.h"
+#include <unistd.h>
```
That's a UNIX header include; the sleep functionality should be provided by HAL.

jesseli2002 commented on avionics/common/include/tasks/restart_camera.hpp:
```diff
@@ -0,0 +1,24 @@
+#pragma once
```
Turns out header guards are useful in that they give you a hack to mock header files in unit tests, by defining the guard itself - so I'm trying to migrate back to using `#ifndef` header guards; besides it's more standards compliant that way

jesseli2002 commented on avionics/common/include/states/main_descent.h:
```diff
@@ -2,10 +2,14 @@
 #include "state_input_struct.h"
 #include "state_interface.h"
+#include "tasks/restart_camera.hpp"
+#include "scheduler.hpp"
```
It's possible to preregister a task (provide the scheduler with task details but not have it actually schedule the task until later), then manually schedule it later - I'd recommend skimming through the actual RocketScheduler header file; I've tried to keep good documentation for the public API. 
This would make it so you wouldn't have to create the dependancy on Cameras in main descent - just preregister the task in the main avionics file, then schedule the task (which only needs the task ID)

akoen commented:
I could write a unit test for this task, but I suspect it would be contrived given that it's functionality depends on manipulating the scheduler while the rocket is running. Do you think it makes sense to do so, or should testing be reserved for full flight tests?

jesseli2002 commented on avionics/common/include/states/landed.h:
```diff
@@ -30,6 +30,11 @@ class Landed : public IState {
+
+        // REVIEW(akoen) disabledRepeat() is more performant, but may cause
+        // errors if the camera is turned on immediately before turning off
+        // again. Test this.
```
`unschedule()` is probably acceptable here; at this point the rocket has landed and there's nothing else for it to do. It's also not that more performant - there's like 4 tasks it has to search through.

jesseli2002 commented on avionics/common/include/cameras.h:
```diff
@@ -15,7 +15,8 @@
  */
 
-#pragma once
+#ifndef Cameras_H_2D865819C2084936A8B94C592E41FE26
```
Other files have been using all caps for the header guards

jesseli2002 commented on avionics/common/include/states/main_descent.h:
```diff
@@ -36,7 +39,13 @@ class MainDescent : public IState {
+
+        // REVIEW(akoen) For aesthetic purposes, it may be desirable to schedule the
+        // task later so that we get sweet parachute deployment footy.
+        Scheduler::scheduleTaskNow(static_cast<int>(TaskID::RestartCamera));
```
I agree - IIRC there's a method for the scheduler that schedules it some time into the future

jesseli2002 commented on avionics/common/include/tasks/restart_camera.hpp:
```diff
@@ -0,0 +1,36 @@
+class RestartCamera {
+private:
+    Camera &cam_;
+    static bool camera_on;
```
Why does this need to be a static variable?
Also, naming convention - suffix private variables with an underscore `camera_on_`

akoen replied:
I was just mirroring LEDBlinker, which uses a static bool for power status. I've gone ahead and changed it though (I just assumed it was needed).

jesseli2002 commented on avionics/common/include/tasks/restart_camera.hpp:
```diff
@@ -0,0 +1,36 @@
+    }
+private:
+
+    void toggle_power() {
```
naming convention - `togglePower` (yes I'm aware that `stop_record()` and `start_record()` don't follow this)

jesseli2002 commented on avionics/tests/unit_RestartCamera.cpp:
```diff
@@ -0,0 +1,30 @@
+    try {
+        Scheduler::runNext();
+        // REVIEW: restart_camera_task.period always reports 0
+        ASSERT_EQ(Scheduler::get_task(taskID).period, 2000);
```
Should probably use the defined constants in `tasks/restart_camera.hpp`

jesseli2002 commented on avionics/tests/fake_include/scheduler.hpp:
```diff
@@ -0,0 +1,18 @@
+#define SCHEDULER_HPP_027EFDB5B9564B41A3DB637CF47A3960
+// #include "scheduler_clock.hpp"
+#include "RocketSchedule/schedule.hpp"
+#include "RocketScheduler/tests/support/fake_clocks.hpp"
```
I don't really like the idea of directly including this test file from the RocketScheduler library, since it's effectively an implementation detail. I would honestly prefer it if you just copied the file contents over; 

Alternatively, a more correct approach would be to reorganize the RocketScheduler library so that `fake_clocks.hpp` and the other reusable fake clocks are in the RocketScheduler include directory, and throw the fake clock definitions under a namespace

jesseli2002 replied:
This comment applies also to `unit_RestartCamera.cpp` and the CMakeLists.txt edits

jesseli2002 commented on avionics/tests/fake_include/scheduler.hpp:
```diff
@@ -0,0 +1,18 @@
+// #include "scheduler_clock.hpp"
+#include "RocketSchedule/schedule.hpp"
+#include "RocketScheduler/tests/support/fake_clocks.hpp"
+// #include "fake_clocks.hpp"
```
delete commented out include files

jesseli2002 commented on avionics/common/include/states/landed.h:
```diff
@@ -30,6 +30,8 @@ class Landed : public IState {
             Hal::ms(LANDED_POLLING_INTERVAL);
+
+        Scheduler::unschedule(static_cast<int>(TaskID::RestartCamera));
         cam_.stop_record();
```
Might want to put a comment here noting that  stop_record ensures that the camera stops recording, since the state of the RestartCamera task is unknown at this point

jesseli2002 commented on avionics/common/include/tasks/restart_camera.hpp:
```diff
@@ -0,0 +1,36 @@
+            camera_on = true;
+            this_task.period = Scheduler::Duration{CAMERA_RESTART_INTERVAL};
+        }
+    }
```
If you want, you can put the implentation of this class into the .cpp file - that'll slightly reduce compilation time

jesseli2002 reviewed (changes requested):
Mostly nitpicky things.

Commit:
Refactored fake_clocks into include/

Commit:
Indentation

jesseli2002 committed:
nah i like this name better

jesseli2002 committed:
Fix invalidated reference bug

jesseli2002 committed:
Squashed onto dev

jesseli2002 committed:
Add 'avionics/common/lib/RocketScheduler/' from commit '4a4b8bcc021f9311c6a07587eddd62369695edd6'

git-subtree-dir: avionics/common/lib/RocketScheduler
git-subtree-mainline: 2b1d361ebb4ed8c1c68a604b5199db4196e62884
git-subtree-split: 4a4b8bcc021f9311c6a07587eddd62369695edd6
